File: TUG.py
from itertools import imap
from sys import stdin

def knapsack(lis):
	count=[0]*105

	for e in lis:
		if count[e]:
			return 1
		count[e]+=1

	SUM = sum(lis)/2+1
	n = len(lis)

	sumi = [0]*(sum(lis)+1)
	sumi[0] = 1

	sum_so_far = 0

	for i in xrange(n):
		for j in xrange(sum_so_far+lis[i],lis[i]-1,-1):
			if SUM>=j and sumi[j-lis[i]]:
				if sumi[j]:
					return 1
				sumi[j] = 1 

		sum_so_far+=lis[i]

	return 0

def main():
	dstream = imap(int, stdin.read().split())
	for i in xrange(next(dstream)):
		
		n = next(dstream)
		lis = [next(dstream) for _ in xrange(n)]


		if knapsack(lis):
			print "YES"
		else:
			print "NO"
main()

File: TUG.cpp

#include<stdio.h>
bool sum[10000001];
int arr[100001],sum_so_far;
int dp(int s,int N)
{
	sum_so_far=0;
	int i,j,k,t,z;
	j=0;
	k=1;
	s/=2;

	for(i=1;i<=s;i++)
		sum[i]=false;

	sum[0]=true;

	for(z=0;z<N;z++)
	{
		for(i=sum_so_far+arr[z];i>=arr[z];i--)
		{
			if(i<=s and sum[i-arr[z]])
			{
				if(sum[i])
				{
					printf("YES\n");
					return 1;
				}
				else
					sum[i]=true;
			}
		}

		sum_so_far+=arr[z];
	}
	return 0;
}
int main()
{
	freopen("input.txt","r",stdin);
	freopen("output.txt","w",stdout);

int T;
scanf("%d",&T);
while(T--)
{
int N,s=0,f,i;
scanf("%d",&N);
for(i=0;i<N;i++)
{
scanf("%d",&arr[i]);
s+=arr[i];
}
f=dp(s,N);
if(!f)
printf("NO\n");
}
return 0;
}